### ISSUE TITLE
Wish: ssh authentication for registry

### ISSUE THREAD

**Issue by gerdstolpmann** [docker/distribution#1470]
I'm wondering why the registry protocol only provides password authentication so far. For sysadmin tasks we normally use ssh with its pubkey authentication, agent forwarding, and the Kerberos option. Passwords do not fit in well here, and are also a bit anachronistic, and we don't want to deploy passwords on servers. So why not allow to pull/push via ssh? E.g.

```
docker pull ssh:user@server/repo/name
```

The idea is to implement this as ssh subsystem (like sftp), so that it does not interfer with normal use of ssh for command execution. You can actually run any protocol over the bidirectional connection so that the HTTP design can be retained, this looks very much like the Unix domain variant. For ssh you'd need to create an account on the registry host for every user, and in the non-free version of the registry you could connect different rights with these users.

**Comment by stevvooe** [comment]
This can already be done with port forwarding. Just bind the registry to localhost, then forward the port from your engine hosts to the registry host. You'll likely have to mark the registry as insecure.

I can't see us ever building this in by default without defining a file layout. Running http over ssh just doesn't perform well. Another option here would be to define a protocol similar to the `git-receive-pack` binary.

**Comment by gerdstolpmann** [comment]
Well, just tried the port forwarding, and it is not really attractive. For one, as you suspected, it is fairly slow (but would have to do more tests to give numbers, maybe the slow down is acceptable, and ssh is tweakable). The other problem is simply naming. If you e.g. forward localhost:5000 to the remote server, then the pulled images will be prefixed with localhost:5000. If the next user prefers localhost:4000, the prefix will be different. And using different ports can be unavoidable on multi-user systems.

My intention was more datacenter use than my personal affairs. Consistent naming would be a requirement for this. We would at least need a way to enforce a deviating transport, e.g. something like

```
docker pull --transport my-registry:localhost:4567 my-registry/repo
```

and/or an environment variable, meaning that my-registry can be reached somewhere else. (Well, maybe we can also already do that with a tricky proxy configuration, but my thinking is that this would be much easier to use if directly supported by the engine.)

**Comment by Teudimundo** [comment]
I agree, I think  would be really convenient not having to create tunnels and redirect addresses to localhost; in particular when you are working with docker machine and it's not localhost but the docker machine host address and you have to open the ssh tunnel to more than localhost.

**Comment by antonioflaviofr** [comment]
Is there any plans to provide ssh transport (and authentication) for the registry? This is an old request and still relevant.

**Comment by manishtomar** [comment]
@antonioflaviofr There are no plans on providing SSH support.

**Comment by DrPyser** [comment]
I also think this would be a really convenient and simple way(for users) to provide secure authentication without the trouble of password management. Most devs using docker will already have ssh keys setup to authenticate to remote servers probably hosting private registries. If docker were to support this out-of-the box or with minimal additional setup, I think this would improve the docker registry experience for devs and ops team in many contexts.
